### PR TITLE
[FileStream] add tests for device and UNC paths

### PR DESCRIPTION
**Pull request by adamsitnik**
Contributes to #54495

Does the following:

* tests unseekable devices by opening a `FileStream` using handle opened from a path pointing to a named pipe (`\\.\pipe\$pipeName`)
* tests seekable devices  by using `\\.\` and `\\?\` paths that use DeviceID instead of a drive letter (sth like `\\.\Volume{724edb31-eaa5-4728-a4e3-f2474fd34ae2}\Users\x\AppData\Local\Temp\y\z` instead `C:\Users\x\AppData\Local\Temp\y\z`)
* tests UNC paths by creating a local file share and accessing it via `\\localhost\$shareName`. This requires an Admin privillige and does not work on Nano, which does not provide "netapi32.dll"

These tests are going to be failing for now, we need to merge #54483 first

### CODE REVIEW

File: src/libraries/System.IO.FileSystem/tests/FileStream/FileStreamConformanceTests.cs [hunk 1]
```diff
@@ -274,4 +278,145 @@ protected override async Task<StreamPair> CreateConnectedStreamsAsync()
...
+            }
+
+            // instead of:
+            // 'C:\Users\x\AppData\Local\Temp\y\z
+            // we want sth like:
+            // '\\.\Volume{724edb31-eaa5-4728-a4e3-f2474fd34ae2}\Users\x\AppData\Local\Temp\y\z
+            string devicePath = filePath.Replace(drive, volumeNameBuffer.ToString());
+            Assert.StartsWith(@"\\?\", devicePath);
+#if DEBUG
+            devicePath = devicePath.Replace(@"\\?\", @"\\.\"); // we want to test both \\.\ and \\?\
```
Comment (stephentoub) [inline src/libraries/System.IO.FileSystem/tests/FileStream/FileStreamConformanceTests.cs:332]: I assume the comment is referring to testing one in debug and the other in release?

**Review by stephentoub — COMMENTED** [review]
Thanks for adding these.  There are some compilation failures, but overall the additions LGTM.

**Comment by adamsitnik** [comment]
The CI failures are unrelated (Android and System.ServiceProcess* failures from outerloop)

File: src/libraries/System.IO.FileSystem/tests/FileStream/FileStreamConformanceTests.Windows.cs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+            // example: "\\?\Volume{724edb31-eaa5-4728-a4e3-f2474fd34ae2}\"
+            if (!GetVolumeNameForVolumeMountPoint(drive, volumeNameBuffer, volumeNameBuffer.Capacity))
+            {
+                throw new Win32Exception(Marshal.GetLastPInvokeError(), "GetVolumeNameForVolumeMountPoint failed");
+            }
+
+            // instead of:
+            // 'C:\Users\x\AppData\Local\Temp\y\z
+            // we want sth like:
+            // '\\.\Volume{724edb31-eaa5-4728-a4e3-f2474fd34ae2}\Users\x\AppData\Local\Temp\y\z
```
Comment (stephentoub) [inline src/libraries/System.IO.FileSystem/tests/FileStream/FileStreamConformanceTests.Windows.cs:63]: "sth"?

File: src/libraries/System.IO.FileSystem/tests/FileStream/FileStreamConformanceTests.Windows.cs [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        }
+
+        [DllImport(Interop.Libraries.Kernel32, EntryPoint = "GetVolumeNameForVolumeMountPointW", CharSet = CharSet.Unicode, BestFitMapping = false, SetLastError = true)]
+        private static extern bool GetVolumeNameForVolumeMountPoint(string volumeName, StringBuilder uniqueVolumeName, int uniqueNameBufferCapacity);
+    }
+
+    [PlatformSpecific(TestPlatforms.Windows)] // the test setup is Windows-specifc
+    [Collection("NoParallelTests")] // don't run in parallel, as file sharing logic is not thread-safe
+    [OuterLoop("Requires admin privileges to create a file share")]
+    [ConditionalClass(typeof(UncFilePathFileStreamStandaloneConformanceTests), nameof(UncFilePathFileStreamStandaloneConformanceTests.CanShareFiles))]
```
Comment (stephentoub) [inline src/libraries/System.IO.FileSystem/tests/FileStream/FileStreamConformanceTests.Windows.cs:81]: Do you know whether the condition ends up being evaluated per test or per class? I don't know the answer,  but if it's the former, might want to cache the static property... it looks pretty expensive.

**Review by stephentoub — APPROVED** [review]
Thanks.

**Comment by danmoseley** [comment]
/backport to release/6.0-preview6

in case we want it there.

**Pull request merged**